- refactored branch trigger

jgangemi opened the pull request:
this should help to stabilize this trigger and allow for additional enhancements. i've refactored the filters so they can be chained and more easily tested. 

skipping the first run should also work properly now b/c i'm checking to see if the branch name exists in the local repo as well. prior to this, it was just skipping everything if the check was enabled.

i will have an improved commit message event implementation that will be ready after we sort through any issues here. i also have a better understanding of how things work now, so i should be more useful and ask less (hopefully) questions.

<!-- Reviewable:start -->
---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kostyasha/github-integration-plugin/147)

<!-- Reviewable:end -->

jgangemi commented:
i'm also quite happy to take on adding back in restriction, etc functionality so you can keep focused on any pull request issues.

KostyaSha commented on github-pullrequest-plugin/src/main/java/com/github/kostyasha/github/integration/branch/GitHubBranchTrigger.java:
```diff
@@ -31,33 +56,13 @@
+import hudson.model.Job;
+import hudson.triggers.Trigger;
+import jenkins.model.ParameterizedJobMixIn;
+import jenkins.triggers.SCMTriggerItem;
```
no reorder.. i'm using IDEA defaults

jgangemi commented:
only `0.69%` 😞

KostyaSha commented on github-pullrequest-plugin/src/main/java/com/github/kostyasha/github/integration/branch/GitHubBranchRepository.java:
```diff
@@ -53,6 +53,14 @@ public GitHubBranchRepository(GHRepository remoteRepository) {
         return branches;
     }
 
+    public void addBranch(GitHubBranch branch) {
```
useless

jgangemi replied:
why? why should i have to access the internal map to add/get a branch? it exposes internal implementation details.

also - as a delegate method, it makes unit testing easier.

KostyaSha replied:
Just `getBranches().put` like in core `getPublishers()` and etc.

KostyaSha commented:
Don't change method signatures

jgangemi commented:
i assume you mean on the events? what is the point of passing the trigger if it's not used? it seems like a useless coupling?

KostyaSha commented:
if it was passed then it was needed, i have no time now digging into details. And PR and branch must have the same logic.

jgangemi commented:
ok - fair point. i will revert that part.

jgangemi commented:
meh - working on the commit message trigger has exposed a flaw. going to improve test cases.

KostyaSha commented on .gitignore:
```diff
@@ -81,3 +81,4 @@ local.properties
 Session.vim
 .netrwhist
 *~
+some/
```
what is it?

KostyaSha replied:
ah, seems wrong written test case

KostyaSha commented on github-pullrequest-plugin/src/main/java/com/github/kostyasha/github/integration/branch/GitHubBranchTrigger.java:
```diff
@@ -222,12 +226,17 @@ public void doRun(String branch) {
+            LoggingTaskListenerWrapper logger) {
+        return GitHubBranchFilterChain.filterChain()
+                .with(skipFirstBuildFilter(skipFirstRun, localRepository, logger));
+    }
```
why it needed?

KostyaSha commented on github-pullrequest-plugin/src/main/java/com/github/kostyasha/github/integration/branch/GitHubBranchTrigger.java:
```diff
@@ -222,12 +226,17 @@ public void doRun(String branch) {
     private List<GitHubBranchCause> readyToBuildCauses(GitHubBranchRepository localRepository,
-                                                       LoggingTaskListenerWrapper listener,
-                                                       @Nullable String branch) {
+            LoggingTaskListenerWrapper listener, @Nullable String branch) {
```
useless change

KostyaSha commented on github-pullrequest-plugin/src/main/java/com/github/kostyasha/github/integration/branch/GitHubBranchTrigger.java:
```diff
@@ -288,6 +286,26 @@ public void doRun(String branch) {
+            GitHubBranchRepository localRepository, LoggingTaskListenerWrapper listener) {
+        List<GitHubBranchCause> causes = remoteBranches.stream()
+                // TODO: update user whitelist filter
+                .filter(createFilterChain(localRepository, listener))
```
unclear method name, gives no idea for what it needed

KostyaSha commented on github-pullrequest-plugin/src/main/java/com/github/kostyasha/github/integration/branch/GitHubBranchTrigger.java:
```diff
@@ -288,6 +286,26 @@ public void doRun(String branch) {
+        long count = remoteBranches.stream()
+                .map(branchRepoUpdater(localRepository))
+                .count();
+        LOG.trace("Updated local branch details with [{}] repositories", count);
```
s/with/for/ i think

KostyaSha commented on github-pullrequest-plugin/src/main/java/com/github/kostyasha/github/integration/branch/events/GitHubBranchEvent.java:
```diff
@@ -39,15 +38,6 @@ public GitHubBranchCause check(
-    /**
-     * Check that is used for lightweight hooks (pure GitHub hooks).
-     */
-    public GitHubBranchCause checkHook(GitHubBranchTrigger githubTrigger,
```
don't remove now

KostyaSha commented on github-pullrequest-plugin/src/main/java/com/github/kostyasha/github/integration/branch/trigger/GitHubBranchFilterChain.java:
```diff
@@ -0,0 +1,30 @@
+
+    @Override
+    public boolean test(GHBranch buildable) {
+        return filters.stream().filter(filter -> filter.test(buildable)).findAny().isPresent();
```
add new lines for calls because it looks messy

KostyaSha commented on github-pullrequest-plugin/src/main/java/com/github/kostyasha/github/integration/branch/trigger/GitHubBranchFilterChain.java:
```diff
@@ -0,0 +1,30 @@
+        return filters.stream().filter(filter -> filter.test(buildable)).findAny().isPresent();
+    }
+
+    public GitHubBranchFilterChain with(Predicate<GHBranch> filter) {
```
`withFilter` then i suppose

KostyaSha commented:
Why you deleted/created functions instead of only changing inheritance?

KostyaSha commented:
@jgangemi better do only small single feature changes and it will be faster to review and merge them.

jgangemi commented:
yes - i reverted back the class name changes, etc and will try to keep this as small as possible. i refactored the filters/ui for branches so they work like the triggers.

here is a screen shot. i wasn't sure how to migrate the 'skip first run' option, but given the branch stuff is experimental, may that doesn't matter. 

<img width="1448" alt="screen shot 2016-10-31 at 9 56 06 am" src="https://cloud.githubusercontent.com/assets/1831839/19856840/a3cde534-9f51-11e6-9106-82753f6e2b11.png">

i should have a new PR for this in a few hours.

KostyaSha commented:
skipFirstRun was just checkbox, you don't need it in filters.... It just a feature that influence on trigger behaviour. If you do some WIP please PR because i need some the same things to do and will be ok to comment and help with something.

KostyaSha commented:
Also `run` is preferred terminology instead `build` usually you have 1 build per ~10 other jobs. And trigger operates on Run interface.

jgangemi commented:
but it still acts like a filter which is why i included it w/ the list. if you disagree i'll revert. 

i'll open a new PR w/ the clean up from your other comments.

KostyaSha commented:
SkipFirstRun is pain for declarative config. It like a flip bit that reverts as soon as trigger firstly runs. job-dsl configuration could be applied multiple times and you can't use it at all. So imho better keep it separate because (and in the same way as done in PR trigger to not have a mess with different impls).

KostyaSha commented:
For branch, and terms i use it close to github API. Branch belong to repo and repo belong to user or  ORG. 
So normally it should have ability to do include/exclude (reverse logic) for any of this criteria. But i think it could be done as separate FilterCretierias, looks like Events logic was good designed for extensibility :)

KostyaSha commented:
Also there is WIP work on having oauth support for trigger and work without global api+token settings. The idea to have the same as Travis simple flow...

KostyaSha commented:
And it would definitely require some APIs from plugin but i can't design them ahead... So what works on freestyle/workflow job is fine for now.

jgangemi commented:
ok - i will revert the skip check...

as part of that, this is the logic for the skip check which seems wrong to me
```
if (skipFirstRun) {
    logger.info("Skipping first run for branch '{}'", remoteBranch.getName());
    return false;
}
```
shouldn't that check also look and see if the branch is in the local repository and if it's not, that means it really is the first run?

jgangemi commented:
replaced by #149
